news/2020-04-01-How-to-configure-wildfly-s2i-by-cli-operations.adoc

```diff
@@ -0,0 +1,446 @@
+* `idle-timeout-minutes`: This attribute specifies the maximum time, in minutes, a connection may be idle before being closed. As idle connections are closed, the number of connections in the pool will shrink down to the specified minimum.
+
+Since we are going to supply our **postconfigure.sh** file in our application Git repository, we will use **install.sh** script to copy this file to the place expected by the WildFly S2I image so it gets executed when the server is launched.
+
+Let us examine the content of our files. First, the https://github.com/yersan/jaxrs-postgresql-demo/blob/cli-extensions/s2i-config/install.sh[install.sh] file:
+
+[source, bash]
+----
+#!/usr/bin/env bash
+
+injected_dir=$1
+
+echo "Copy ${injected_dir}/extensions/postconfigure.sh to ${JBOSS_HOME}/extensions/"
+
+mkdir -p "${JBOSS_HOME}/extensions/"
+cp "${injected_dir}/extensions/postconfigure.sh" "${JBOSS_HOME}/extensions/"
+----
+
+It content is pretty simple; creates the "${JBOSS_HOME}/extensions/" if it does not exist yet, and copies our **postconfigure.sh** script.
+
+Now let us look at the content of our https://github.com/yersan/jaxrs-postgresql-demo/blob/cli-extensions/s2i-config/extensions/postconfigure.sh[postconfigure.sh] script:
+
+[source, bash]
+----
+#!/usr/bin/env bash
+
+echo "Appending CLI operations to ${CLI_SCRIPT_FILE}"
+
+echo "
+  /subsystem=datasources/data-source=database_server_postgresql-DSDEMO:write-attribute(name=pool-use-strict-min, value=true)
+  /subsystem=datasources/data-source=database_server_postgresql-DSDEMO:write-attribute(name=idle-timeout-minutes, value=5)
+" >> "${CLI_SCRIPT_FILE}"
+----
+
+We can append CLI operations to the final CLI script used by the WildFly image. You can access to this file through the environment variable `CLI_SCRIPT_FILE` which is available in this script environment.
+
+The management operations executed in this script assume there is already a datasource named `database_server_postgresql-DSDEMO`. This datasource will be created and configured by using the standard environment variables.
+
+=== Test the application
+
+Assuming your database server is already configured as in our previous example, let us now create our Openshift application using this time the cli-extensions branch and by passing in the environment variables that configures our data source:
+
+[source, bash]
+----
+$ oc new-app --name wildfly-app \
+        https://github.com/yersan/jaxrs-postgresql-demo.git#cli-extensions  \
+        --image-stream=wildfly \
+        --env DB_SERVICE_PREFIX_MAPPING="database-server-postgresql=DSDEMO" \
+        --env DSDEMO_JNDI="java:/jboss/datasources/PostgreSQLDS" \
+        --env DSDEMO_USERNAME="postgre" \
+        --env DSDEMO_PASSWORD="admin" \
+        --env DSDEMO_DATABASE="demodb" \
+        --env DSDEMO_DRIVER="postgresql" \
+        --env DATABASE_SERVER_POSTGRESQL_SERVICE_HOST="database-server" \
```

Comment:
I'm not still convinced by the uses of these environment variables. It looks like we have a bug in the way we use DB_SERVICE_PREFIX_MAPPING and in our documentation.

Reply:
I am not 100% convinced this is the correct way, specify `DATABASE_SERVER_POSTGRESQL_SERVICE_HOST` and `DATABASE_SERVER_POSTGRESQL_SERVICE_PORT` should not be necessary. Our scripts should be adapted to remove this need, those variables should be retrieved automatically from the pod environment; in Openshift, when a service is exposed, in this case the database-server service, the following variables are available at runtime to any other pod in the cluster (service-name_PORT and service-name_HOST):
`DATABASE_SERVER_SERVICE_HOST`
`DATABASE_SERVER_SERVICE_PORT`

However, we are still "forced" to pass in the other variants (service-name_databstype_PORT and service-name_databstype_HOST).

They are there just because the scripts where originally prepared upon the existence of the templates that used posgresql and mysql database. Those templates configured the services in the template, which exposed those variables, but we are no longer supporting them, the persisted templates where removed, however we are still using them, I guess for backward compatibility.

It would be better to get those values from the services exposed by the data base pod instead of to have to specify them manually because we are no longer using those removed templates. I suppose this change would be a feature request, so by now, I would be fine if we post this article as it is.

@bstansberry I'm not planning to apply more changes to this blog entry, I'm just waiting for its review and feedback.